File: app/src/main/java/dev/blackcat/minauta/net/JNautaConnection.java
package dev.blackcat.minauta.net;

import dev.blackcat.jnauta.net.Authentication;
import dev.blackcat.jnauta.net.AuthenticationResponseParser;
import dev.blackcat.jnauta.net.ConnectionBuilder;
import dev.blackcat.minauta.sync.AvailableTimeThread;
import dev.blackcat.minauta.sync.LoginThread;
import dev.blackcat.minauta.sync.LogoutThread;
import dev.blackcat.minauta.sync.SyncThread;
import dev.blackcat.minauta.data.Account;
import dev.blackcat.minauta.data.Session;

public class JNautaConnection implements Connection
{

    @Override
    public LoginResult login(Account account)
    {
        LoginThread thread = new LoginThread(account);
        AuthenticationResponseParser.LoginResult loginResult = thread.execute();

        LoginResult result = new LoginResult();
        if (loginResult == null)
            result.state = State.ERROR;
        else if (loginResult.alreadyConnected)
            result.state = State.ALREADY_CONNECTED;
        else if (loginResult.noMoney)
            result.state = State.NO_MONEY;
        else if (loginResult.badPassword)
            result.state = State.INCORRECT_PASSWORD;
        else if (loginResult.badUsername)
            result.state = State.UNKNOWN_USERNAME;
        else if (loginResult.isGoogle)
            result.state = State.IS_GOOGLE;
        else if (loginResult.session.equals("") || loginResult.timeParams.equals(""))
            result.state = State.ERROR;
        else
        {
            Session session = new Session();
            session.setLogoutParams(loginResult.session);
            session.setTimeParams(loginResult.timeParams);
            result.session = session;
            result.state = State.OK;
        }
        return result;
    }

    @Override
    public LogoutResult logout(Account account)
    {
        LogoutThread thread = new LogoutThread(account);
        Boolean logoutResult = thread.execute();

        LogoutResult result = new LogoutResult();
        result.state = State.OK;
        if (logoutResult == null || !logoutResult)
            result.state = State.ERROR;
        return result;
    }

    @Override
    public AvailableTimeResult getAvailableTime(Account account)
    {
        AvailableTimeThread thread = new AvailableTimeThread(account);
        String time = thread.execute();

        AvailableTimeResult result = new AvailableTimeResult();
        if (time == null)
            result.state = State.ERROR;
        else
        {
            result.state = State.OK;
            result.availableTime = time;
        }
        return result;
    }
}

File: app/src/main/java/dev/blackcat/minauta/net/JNautaConnection.kt
package dev.blackcat.minauta.net

import dev.blackcat.jnauta.net.Authentication
import dev.blackcat.jnauta.net.ConnectionBuilder
import dev.blackcat.minauta.data.Account
import dev.blackcat.minauta.data.Session
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.*

class JNautaConnection : Connection {

    override suspend fun login(account: Account): Connection.LoginResult {
        val authentication = Authentication(ConnectionBuilder.Method.OK_HTTP, account.username, account.password, null)
        val loginResult = authentication.login()
        val result = Connection.LoginResult()
        if (loginResult == null)
            result.state = Connection.State.ERROR
        else if (loginResult.alreadyConnected)
            result.state = Connection.State.ALREADY_CONNECTED
        else if (loginResult.noMoney)
            result.state = Connection.State.NO_MONEY
        else if (loginResult.badPassword)
            result.state = Connection.State.INCORRECT_PASSWORD
        else if (loginResult.badUsername)
            result.state = Connection.State.UNKNOWN_USERNAME
        else if (loginResult.isGoogle)
            result.state = Connection.State.IS_GOOGLE
        else if (loginResult.paramString == "")
            result.state = Connection.State.ERROR
        else {
            val session = Session()
            session.loginParams = loginResult.paramString
            session.startTime = Calendar.getInstance().timeInMillis
            result.session = session
            result.state = Connection.State.OK
        }
        return result
    }

    override suspend fun logout(account: Account, session: Session): Connection.LogoutResult {
        val authentication = Authentication(ConnectionBuilder.Method.OK_HTTP, account.username, account.password, null)
        val logoutResult = authentication.logout(session.loginParams)

        val result = Connection.LogoutResult()
        result.session = session
        result.state = if (logoutResult) Connection.State.OK else Connection.State.ERROR
        return result
    }

    override suspend fun getAvailableTime(account: Account, session: Session): Connection.AvailableTimeResult {
        val authentication = Authentication(ConnectionBuilder.Method.OK_HTTP, account.username, account.password, null)
        val time = authentication.getAvailableTime(session!!.loginParams)

        val result = Connection.AvailableTimeResult()
        result.state = if (time != null) Connection.State.OK else Connection.State.ERROR
        result.availableTime = time
        return result
    }

}
